go/consensus/pricediscovery/pricediscovery.go

```diff
@@ -0,0 +1,161 @@
+// Package pricediscovery implements dynamic price discovery.
+package pricediscovery
+
+import (
+	"context"
+	"sync"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/quantity"
+	consensus "github.com/oasisprotocol/oasis-core/go/consensus/api"
+)
+
+const (
+	// defaultWindowSize is the default number of recent blocks to use for calculating min gas
+	// price.
+	//
+	// NOTE: Code assumes that this is relatively small.
+	defaultWindowSize uint64 = 6
+
+	// defaultFullBlockThreshold is the default percentage of block used gas over which a block
+	// should be considered full.
+	defaultFullBlockThreshold float64 = 0.5
+)
+
+type priceDiscovery struct {
+	mu sync.RWMutex
+
+	fallbackGasPrice *quantity.Quantity
+	minGasPrice      *quantity.Quantity
+	computedGasPrice *quantity.Quantity
+
+	// blockPrices is a rolling-array containing minimum transaction prices for last up to
+	// `windowSize` blocks.
+	blockPrices []*quantity.Quantity
+	// tracks the current index of the blockPrices rolling array.
+	blockPricesCurrentIdx int
+
+	windowSize         uint64
+	fullBlockThreshold float64
+
+	client consensus.ClientBackend
+
+	logger *logging.Logger
+}
+
+// GasPrice implements consensus.PriceDiscovery.
+func (pd *priceDiscovery) GasPrice(_ context.Context) (*quantity.Quantity, error) {
+	pd.mu.RLock()
+	defer pd.mu.RUnlock()
+
+	// Choose the maximum of (fallback, min, computed) gas prices.
+	gasPrice := pd.fallbackGasPrice.Clone()
+	if pd.computedGasPrice != nil && pd.computedGasPrice.Cmp(gasPrice) > 0 {
+		gasPrice = pd.computedGasPrice.Clone()
+	}
+	if pd.minGasPrice.Cmp(gasPrice) > 0 {
+		gasPrice = pd.minGasPrice.Clone()
+	}
+	return gasPrice, nil
+}
+
+func (pd *priceDiscovery) refreshMinGasPrice(ctx context.Context) {
+	mgp, err := pd.client.MinGasPrice(ctx)
+	if err != nil {
+		pd.logger.Warn("failed to refresh minimum gas price",
+			"err", err,
+		)
+		return
+	}
+
+	pd.mu.Lock()
+	defer pd.mu.Unlock()
+
+	pd.minGasPrice = mgp
+}
+
+func (pd *priceDiscovery) processBlock(_ context.Context, _ *consensus.Block) {
+	// Currently transactions are not ordered by price, so track price as zero. After the mempool
+	// is refactored, change this to properly compute the median gas price.
+	//
+	// We should also make sure to add some margin over the median in case of full blocks.
+	pd.trackPrice(quantity.NewQuantity())
+}
+
+func (pd *priceDiscovery) trackPrice(price *quantity.Quantity) {
+	// One item always gets added added to the prices array.
+	// Bump the current index for next iteration.
+	defer func() {
+		pd.blockPricesCurrentIdx = (pd.blockPricesCurrentIdx + 1) % int(pd.windowSize)
```

Comment:
```suggestion
		pd.blockPricesCurrentIdx = (pd.blockPricesCurrentIdx + 1) % int(len(pd.blockPrices))
```
You can get rid of this field.